### PR TITLE
Only add primary column when needed.

### PR DESCRIPTION
**Pull request by peterwilsoncc**
https://core.trac.wordpress.org/ticket/34564

### CODE REVIEW

File: src/wp-admin/includes/class-wp-list-table.php [hunk 1]
```diff
@@ -1108,20 +1108,29 @@ protected function get_primary_column_name() {
...
 			/*
 			 * Backward compatibility for `$_column_headers` format prior to WordPress 4.3.
 			 *
 			 * In WordPress 4.3 the primary column name was added as a fourth item in the
 			 * column headers property. This ensures the primary column name is included
 			 * in plugins setting the property directly in the three item format.
 			 */
+			if ( 4 === count( $this->_column_headers ) ) {
+				return $this->_column_headers;
+			}
```
Comment (spacedmonkey) [inline src/wp-admin/includes/class-wp-list-table.php:1124]: What is happening here?

**Review by spacedmonkey — APPROVED** [review]

**Review by mukeshpanchal27 — APPROVED** [review]
Looks great!

File: tests/phpunit/tests/admin/wpListTable.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+		$column_headers = new ReflectionProperty( $list_table, '_column_headers' );
+		$column_headers->setAccessible( true );
+		$column_headers->setValue( $list_table, $headers );
+
+		$column_info = new ReflectionMethod( $list_table, 'get_column_info' );
+		$column_info->setAccessible( true );
+
+		$this->assertSame( $expected, $column_info->invoke( $list_table ) );
+		$this->assertSame( $expected_hook_count, $hook->get_call_count() );
```
Comment (costdev) [inline tests/phpunit/tests/admin/wpListTable.php]:
```suggestion
		$this->assertSame( $expected, $column_info->invoke( $list_table ), 'The actual columns did not match the expected columns' );
		$this->assertSame( $expected_hook_count, $hook->get_call_count(), 'The hook was not called the expected number of times' );
```

File: tests/phpunit/tests/admin/wpListTable.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	 *
+	 * @ticket 34564
+	 *
+	 * @dataProvider data_should_only_add_primary_column_when_needed
+	 *
+	 * @covers WP_List_Table::get_column_info
+	 *
+	 * @param string $list_class The name of the WP_List_Table child class.
+	 * @param array  $headers    A list of column headers.
+	 * @param array  $expected   The expected column headers.
```
Comment (costdev) [inline tests/phpunit/tests/admin/wpListTable.php]:
```suggestion
	 * @param string $list_class          The name of the WP_List_Table child class.
	 * @param array  $headers             A list of column headers.
	 * @param array  $expected            The expected column headers.
	 * @param int    $expected_hook_count The expected number of times the hook is called.
```

**Review by costdev — CHANGES_REQUESTED** [review]

File: tests/phpunit/tests/admin/wpListTable.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	 * @param int    $expected_hook_count The expected number of times the hook is called.
+	 */
+	public function test_should_only_add_primary_column_when_needed( $list_class, $headers, $expected, $expected_hook_count ) {
+		$hook = new MockAction();
+		add_filter( 'list_table_primary_column', array( $hook, 'filter' ) );
+
+		// This is reset between tests.
+		$GLOBALS['hook_suffix'] = 'my-hook';
+
+		$list_table = _get_list_table( $list_class );
```
Comment (costdev) [inline tests/phpunit/tests/admin/wpListTable.php]:
```suggestion
		$list_table = _get_list_table( $list_class, array( 'screen' => 'test-screen' ) );
```
Passing a `screen` value in the args removes the need to set the `hook_suffix` global.

**Review by costdev — APPROVED** [review]

**Comment by peterwilsoncc** [comment]
Merged https://core.trac.wordpress.org/changeset/53868 / 11cc06cd22

**Pull request closed without merging**